### DISCUSSION TITLE
Search highlight disappears due to focus leaving

### DISCUSSION THREAD

**Discussion by xianyuIC** [NeoVintageous/NeoVintageous#912] [Q&A] [answered]
When I enter `/string` and press the Enter key key, all matching strings will be highlighted and the color is very beautiful. But when I clicked on another place with my mouse, such as my terminal console, and then returned to the sublime, the search highlight disappeared.
I found something in help nv:
```
'vintagous_default_mode' 
Enter insert mode when opening or switching views or when the
window receives focus and is not in visual mode i.e.visual mode
selections are retained when the window loses focus.
```
Is there a setting that can simultaneously meet the following conditions:
1. When I open a new file, it is in visual mode.
2. When I search for the highlight, the mouse leaves the sublime and then returns to the sublime, the search highlight still exists.
3. When I was in insert mode, the mouse left the sublime and then returned to the sublime, still in insert mode.

**Comment by gerardroche** [comment, marked as answer]
2 and 3 are known issues. They're in the roadmap "Rework mode handling to work like Vim". It requires a massive change to how we handle the mode. 

The default visual mode (1) I think might be doable though I would expect some edge case issue that might be difficult to fix. You can try in the branch `default-visual-mode`:

Add default visual mode setting

Menu → Preferences → Settings

```js
"vintageous_default_mode": "visual",
```

**Reply by xianyuIC** [reply]
Even if I don't add 'vintageous_default_mode': 'visual', it actually defaults to visual mode when opening a new file. Sorry, this should be due to my lack of rigor in the above question.

Thank you for your explanation, Best wishes.

**Comment by gerardroche** [comment]
Yes. Visual selections are cleared when switching views. You'll notice that if you make a selection and switch the view it clears the selection in the view you are leaving i.e. the inactive selections. This mirrors Vim behaviour. The reasoning is that a visual selection will only be visible in an active view.

If you make a selection and open the command line `:` and then press escape to close the command line, the visual selection should not be cleared because you didn't switch views i.e. if Sublime loses focus it should not clear the active view visual selection.

This is how the search highlighting should work, but currently the search highlighting is lost when sublime loses focus. I might be able to resolve this issue.

**Comment by gerardroche** [comment]
I just noticed that in Vim the escape key doesn't clear search highlighting. I wonder if that's the default behaviour in vim or if it's controlled by an option.

In Vim, it seems you have to explicitly clear search highlighting via something like `:nohlsearch`. I have this mapped to `CTRL-l` for easy access. I use it all the time. 

```
noremap <C-l> :nohlsearch<CR>
```

Currently neovintageous the search highligting is cleared when pressing escape.

**Reply by xianyuIC** [reply]
In Vim, entering `:noh` can also clear the search highlight, which is also useful in NeoVintagous. NeoVintageous did a great job at this point.

**Comment by gerardroche** [comment]
In the next version you will be able set `vintageous_auto_nohlsearch_on_normal_enter` to false to keep highlighting until you explicitly clear it with `:noh`. I've kept the default to true for the moment because that's the current behaviour, but in a later version I will change the default so that it works like Vim by default.

**Reply by xianyuIC** [reply]
I downloaded the latest version and tried it out according to your description. It can keep highlighted. Thank you very much!

**Comment by gerardroche** [comment]
I can't reproduce the issue. It was big update. You might need to restart Sublime and check the console log for any errors (Menu → View → Show Console).

I'm in the process of write a release blog post, there is this draft section:

## Clearing of Search Highlighting with :nohlsearch

Control the behaviour of search highlighting clearance upon entering Normal mode with a new setting.

Currently, when you enter Normal mode by pressing `<Esc>`, `CTRL-[`, or `CTRL-c`, active search highlighting is automatically cleared. This deviates from Vim's behaviour, where search highlighting is only cleared with the ex command `:nohlsearch` or `:noh`.

To preserve Vim-like behaviour and prevent automatic clearing of highlighting upon entering Normal mode:

1. Open the Command Palette with `Ctrl+Shift+P` (or `Cmd+Shift+P` on macOS).

2. Type "Preferences: NeoVintageous Settings" and confirm with `Enter`.

3. Add the following JSON configuration:

   ```json
   {
      "vintageous_auto_nohlsearch_on_normal_enter": false
   }
   ```

Apply this configuration to maintain search highlighting until explicitly cleared using `:nohlsearch`.

To clear search highlighting directly, use the command:

```vim
:noh[lsearch]
```

For instance, you can map a key combination to clear search highlighting:

1. Launch the Command Palette: `NeoVintageous: Open neovintageousrc file`.

2. Add the following (customize `<C-l>` as desired):

  ```vim
  noremap <C-l> :nohlsearch<CR>
  ```

3. Ensure changes take effect by reloading the `neovintageousrc` file: `NeoVintageous: Reload neovintageous file`.

**Reply by xianyuIC** [reply]
Oh, I may not have expressed myself clearly. Following your prompt, I succeeded: my sublime can be highlighted and won't lose its highlight due to losing focus.
As for the draft you posted below, it's very good, so that more people can know about this major update.
